### PR TITLE
Remove all explicit usage of fmtlib

### LINKED ISSUE #1710: Prepare for CUDA Python new documentation layout
https://github.com/rapidsai/rmm/blob/1024a1250cfde7e93d26dc6d5e063e84c4a39824/python/rmm/docs/conf.py#L200

Once this PR https://github.com/NVIDIA/cuda-python/pull/183 is merged I expect that RMM doc build will break again. 

Based on [this search team](https://github.com/search?q=nvidia.github.io%2Fcuda-python+language%3APython+path%3A*%2Fconf.py+-is%3Afork&type=code) RMM is the only consumer of CUDA Python's `objects.inv`, so wanted to inform you in advance 🙂

### LINKED ISSUE #1717: [FEA] Remove usage of fmt
Currently rmm has some usage of the fmt library. While fmt is convenient, it causes us significant headaches due to the need to keep our fmt dependency in sync with conda-forge, and because fmt's support for header-only usage makes it very easy to accidentally leak its symbols into public DSO symbol tables (which can cause serious runtime issues if multiple libraries do this with incompatible symbols, e.g. with wheels). Most of RAPIDS's fmt usage comes transitively via spdlog, so not exposing spdlog symbols in public APIs (see https://github.com/rapidsai/build-planning/issues/104) is sufficient to remove a public fmt dependency, but rmm is a special case because it also has direct usage of fmt. Fortunately, fmt can be replaced relatively easily with `std::snprintf` in the short term, and with std::format in the future [when we upgrade to building with C++20](https://github.com/rapidsai/build-planning/issues/113). We should replace all existing use cases of fmt so that we can remove our public reliance on the library.

### PR DESCRIPTION
**Pull request by harrism**
## Description
Fixes #1717
Also fixes #1710 in 5330063

I have replaced fmt-style format string placeholders (`"... {} ..."`) with printf-style placeholders by adding a function `rmm::detail::formatted_log()`, which I modified from @vyasr 's #1722.

~The only remaining mention of fmt is in CMakeLists.txt. Do we still need to explicitly fetch fmt?~
Removed.

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rmm/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

File: include/rmm/mr/device/tracking_resource_adaptor.hpp
Comment (wence-) [inline include/rmm/mr/device/tracking_resource_adaptor.hpp:240]:
```suggestion
          "Deallocating a pointer that was not tracked. Ptr: %p [%dB], Current Num. Allocations: "
```

File: include/rmm/mr/device/tracking_resource_adaptor.hpp
Comment (wence-) [inline include/rmm/mr/device/tracking_resource_adaptor.hpp:254]: *issue (perhaps?)*: throughout, `%d` is the format string for _signed integers_. But these are `std::size_t` values, so we need a `%uz` here

File: include/rmm/detail/format.hpp
Comment (wence-) [inline include/rmm/detail/format.hpp:36]: *nit*: add `@throw` description.

File: include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp
Comment (wence-) [inline include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp:203]: *nit*: `format_stream` produces a decimal representation. Is there a reason we couldn't just use:

```
RMM_LOG_TRACE("[A][stream %p][%dB], stream.value(), size);
```
?

**Review by wence- — CHANGES_REQUESTED** [review]
One definite issue in the snprintf calls, the rest is all minor nits, I think

File: include/rmm/detail/format.hpp
Comment (harrism) [reply include/rmm/detail/format.hpp:45]: Good catch. I copy pasted this from @vyasr and I noticed it may have come from here: https://stackoverflow.com/a/26221725/749748 However that answer just used `int`.  I'm not sure why a cast is needed here at all.

File: include/rmm/mr/device/tracking_resource_adaptor.hpp
Comment (harrism) [reply include/rmm/mr/device/tracking_resource_adaptor.hpp:240]: Oops.

File: include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp
Comment (harrism) [reply include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp:203]: `%p` outputs zero as the string "(nil)", which I found out by failing `logger_tests.cpp`.  I don't think we want the null stream to be printed as "(nil)".

Comment (harrism) [reply include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp:203]: Appears that %p formatting is implementation defined. https://en.cppreference.com/w/c/io/fprintf

Comment (wence-) [reply include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp:203]: Ah, that is a good reason not to use `%p`

File: include/rmm/detail/format.hpp
Comment (wence-) [reply include/rmm/detail/format.hpp:45]: > I'm not sure why a cast is needed here at all.

The `size` argument to `snprintf` is a `std::size_t`, so in some scenarios a compiler _might_ warn about the sign conversion (e.g. clang with `-Wsign-conversion`).

Comment (harrism) [reply include/rmm/detail/format.hpp:45]: OK. I tested on compiler explorer with the latest clang and GCC and get no warnings when just using `auto` with no cast.

File: include/rmm/mr/device/tracking_resource_adaptor.hpp
Comment (harrism) [reply include/rmm/mr/device/tracking_resource_adaptor.hpp:254]: Thanks. Note that in my reading, technically this should be `%zu` for broadest compatibility. All modifiers including the length modifier (`z`) go before the conversion format specifier (`u`).

File: include/rmm/detail/format.hpp
Comment (harrism) [reply include/rmm/detail/format.hpp:39]: Added.

File: include/rmm/mr/device/tracking_resource_adaptor.hpp
Comment (harrism) [reply include/rmm/mr/device/tracking_resource_adaptor.hpp]: Removed all instances of %d for more appropriate formatters.

File: include/rmm/detail/format.hpp
Comment (wence-) [reply include/rmm/detail/format.hpp]: Ha, neither clang nor gcc put `-Wsign-conversion` in `-Wall`.

**Review by wence- — APPROVED** [review]
Thanks Mark

File: include/rmm/detail/format.hpp
Comment (harrism) [reply include/rmm/detail/format.hpp]: Why would "all" not include all warnings?!

Anyway, your suggestion of ssize_t gives the same warning as int when `Wsign-conversion` is specified, naturally. I also think the original conditional was wrong anyway because it was done after adding 1.

Comment (wence-) [reply include/rmm/detail/format.hpp]: > Why would "all" not include all warnings?!

A question for the ages :P.

But yes, mine was also bugged in the same way.

File: include/rmm/detail/format.hpp
Comment (wence-) [inline include/rmm/detail/format.hpp:56]: Argh, sorry @harrism, I missed the +1 here when we're obtaining size. So I think a bunch of these conditions have fencepost errors.

I think we `RMM_EXPECTS(size > 0)` and then `if (size == 1) { return {}; }`

Or, I suppose we can do:

```suggestion
  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-vararg)
  auto size = std::snprintf(nullptr, 0, format.c_str(), args...);
  // size is length of string without trailing '\0'
  RMM_EXPECTS(size >= 0, "Error during formatting.");
  if (size == 0) { return {}; }
  // NOLINTNEXTLINE(modernize-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays)
  std::unique_ptr<char[]> buf(new char[size + 1]);
  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-vararg)
  std::snprintf(buf.get(), size + 1, format.c_str(), args...);
  return {buf.get(), buf.get() + size};
```

WDYT?

Comment (harrism) [reply include/rmm/detail/format.hpp]: I saw this after I committed a similar but DRYer version. :)

Comment (wence-) [reply include/rmm/detail/format.hpp]: Yes, thanks!

**Comment by vyasr** [comment]
It looks like you found a number of the issues with my snprintf implementation that I have since fixed :joy: Hopefully I caught everything, I'll review now and see.

File: include/rmm/detail/format.hpp
Comment (vyasr) [inline include/rmm/detail/format.hpp:83]: Nice I like this approach for streams.

File: tests/mr/device/arena_mr_tests.cpp
Comment (vyasr) [inline tests/mr/device/arena_mr_tests.cpp:577]: Unrelated improvement?

**Review by vyasr — CHANGES_REQUESTED** [review]
Overall looks great. I have a few small suggestions, and I'm blocking merge per discussion of https://github.com/rapidsai/rmm/pull/1722#discussion_r1841290495, but can approve once we resolve either way.

File: tests/mr/device/arena_mr_tests.cpp
Comment (harrism) [reply tests/mr/device/arena_mr_tests.cpp:577]: yes fixing a clang-tidy complaint.

File: include/rmm/mr/device/arena_memory_resource.hpp
Comment (harrism) [reply include/rmm/mr/device/arena_memory_resource.hpp:340]: Do you think I should take your "convert_to_c_str" internal function to make it so users don't have to call c_str() everywhere like this?

**Comment by harrism** [comment]
> I have a few small suggestions

@vyasr none of your comments read as suggestions, just comments on differences from your PR. Are there any specific changes you are requesting?

File: include/rmm/mr/device/arena_memory_resource.hpp
Comment (vyasr) [reply include/rmm/mr/device/arena_memory_resource.hpp:340]: Yes I think that is worthwhile. I wasn't a big fan of my `to_string` template solution since that's an implicit contract that we'd need to teach again, but I think supporting std::string transparently is worthwhile and it's not a surprising behavior.

File: benchmarks/utilities/log_parser.hpp
Comment (vyasr) [reply benchmarks/utilities/log_parser.hpp:154]: My main question here is that since `%p` is documented as producing implementation-defined behavior and I haven't tested this on enough different systems, should we be concerned about the potential for differences? Given that CI is passing here I'm comfortable saying no, but I wanted to get your opinion. We could also put together a small snippet on Godbolt to see what it outputs on different platforms.

**Comment by vyasr** [comment]
I resolved the indeterminate threads and left open the two that require nonzero resolution.

File: tests/mr/device/arena_mr_tests.cpp
Comment (harrism) [reply tests/mr/device/arena_mr_tests.cpp:577]: Are you saying you want this moved out of this PR?

File: benchmarks/utilities/log_parser.hpp
Comment (harrism) [reply benchmarks/utilities/log_parser.hpp:154]: Clang and GCC both print "(nil)" on Godbolt. MSVC prints "00000000". 

The nice thing about `%p` is that for non-null pointers you get the `0x` prefix that `%llx` doesn't have: 
code: 
```
int x = 10;
printf("p null: %p, non-null %p\n", nullptr, &x);
printf("llx null: %llx, non-null %llx\n", nullptr, &x);
```
output: 
```
p null: (nil), non-null 0x7ffdc74cc08c
llx null: 0, non-null 7ffdc74cc08c
```

Note that either way I think we should leave this change to `parse_pointer` in place to avoid the runtime errors caused by calling `stoll` on "(nil)".

Comment (vyasr) [reply benchmarks/utilities/log_parser.hpp:154]: I preferred %p formatting when I was testing in #1722 for the hex prefix as well, I was mostly concerned about whether `nil` was a reliable output. It seems like it's standard enough as long as clang and gcc both do it. I'm not concerned about MSVC. So I say let's keep %p, then.

File: include/rmm/mr/device/arena_memory_resource.hpp
Comment (harrism) [reply include/rmm/mr/device/arena_memory_resource.hpp]: Done.

File: benchmarks/utilities/log_parser.hpp
Comment (harrism) [reply benchmarks/utilities/log_parser.hpp:154]: I agree.

**Pull request merged**
